lib/connections/limiter.go

```diff
@@ -12,53 +12,197 @@ import (
+
+			lim.setDeviceLimiters(v.DeviceID, writeLimiter, readLimiter)
+		} else if !okR || !okW {
+			// One of the read/write limiters is not present while the
+			// corresponding write/read one exists. Something has gone wrong.
+			panic("broken symmetry in device read/write limiters")
+		}
+
+		// limiters for this device are created so we can store previous rates for logging
+		previousReadLimit := readLimiter.Limit()
+		previousWriteLimit := writeLimiter.Limit()
+
+		l.Debugf("okR: %t, okW: %t, prevWriteLim: %d, prevReadLim: %d", okR, okW, previousReadLimit, previousWriteLimit)
+
+		lim.setLimitsForDevice(v, readLimiter, writeLimiter)
+
+		// Nothing about this device has changed. Start processing next device
+		if okR && okW &&
+			readLimiter.Limit() == previousReadLimit &&
+			writeLimiter.Limit() == previousWriteLimit {
+			continue
+		}
+
+		readLimitStr := "is unlimited"
+		if v.MaxRecvKbps > 0 {
+			readLimitStr = fmt.Sprintf("limit is %d KiB/s", v.MaxRecvKbps)
+		}
+		writeLimitStr := "is unlimited"
+		if v.MaxSendKbps > 0 {
+			writeLimitStr = fmt.Sprintf("limit is %d KiB/s", v.MaxSendKbps)
+		}
+
+		l.Infof("Device %s send rate %s, receive rate %s", v.DeviceID, readLimitStr, writeLimitStr)
+	}
+
+	// Delete remote devices which were removed in new configuration
+	for deviceID, shouldBeRemoved := range devicesToRemove {
+		l.Debugf("deviceID: %s, shouldBeRemoved: %t", deviceID, shouldBeRemoved)
+		if shouldBeRemoved {
+			lim.deleteDeviceLimiters(deviceID)
+		}
+	}
+
+	l.Debugln("Rebuild of device limiters map finished")
 }
 
-func (lim *limiter) newWriteLimiter(w io.Writer, isLAN bool) io.Writer {
-	return &limitedWriter{writer: w, limiter: lim, isLAN: isLAN}
+// Compare read/write limits in configurations
+func (lim *limiter) checkDeviceLimits(from, to config.Configuration) bool {
+	if len(from.Devices) != len(to.Devices) {
+		return false
+	}
+	// len(from.Devices) == len(to.Devices) so we can do range from.Devices
+	for i := range from.Devices {
+		if from.Devices[i].DeviceID != to.Devices[i].DeviceID {
+			// Something has changed in device configuration
+			return false
+		}
+		// Read/write limits were changed for this device
+		if from.Devices[i].MaxSendKbps != to.Devices[i].MaxSendKbps ||
+			from.Devices[i].MaxRecvKbps != to.Devices[i].MaxRecvKbps {
+			return false
+		}
+	}
+	return true
 }
 
 func (lim *limiter) VerifyConfiguration(from, to config.Configuration) error {
 	return nil
 }
 
 func (lim *limiter) CommitConfiguration(from, to config.Configuration) bool {
+	// to ensure atomic update of configuration
+	lim.mu.Lock()
+	defer lim.mu.Unlock()
+
 	if from.Options.MaxRecvKbps == to.Options.MaxRecvKbps &&
 		from.Options.MaxSendKbps == to.Options.MaxSendKbps &&
-		from.Options.LimitBandwidthInLan == to.Options.LimitBandwidthInLan {
+		from.Options.LimitBandwidthInLan == to.Options.LimitBandwidthInLan &&
+		lim.checkDeviceLimits(from, to) {
 		return true
 	}
```

Comment:
I think this global check is not very useful as global limits are reset even if just a device changed and whether device change is checked again later on. So I would either not check at all and just reset everything or check the different parameters separately.